### DISCUSSION TITLE
How to read compound dataset with complex structure ?

### DISCUSSION THREAD

**Discussion by wlaroche** [LiorBanai/HDF5-CSharp#203] [Q&A]
Hi,

Is it possible to read compound dataset of a complex structure like this, and how ?

```
var fileId = Hdf5.OpenFile("myTrace.h5");
var data = Hdf5.ReadCompounds<ComplexStruct>(fileId, "R::Root/O_MyComplexStruct", "");
```

```
public struct ComplexStruct
{
	public SimpleStruct MySimpleStruct;
	public SimpleStruct[] MySimpleArrayStruct;
	public int myInt;
	public float[] MyArrayFloat;
}

public struct SimpleStruct
{
	public byte A;
	public int B;
	public int C;
}
```

Regards.

**Comment by LiorBanai** [comment]
Hi @wlaroche ,
It should be possible. Can you attach a file for me to check?

**Reply by wlaroche** [reply]
currently it's a little tricky to attach some data, i have just written an example to check if it is possible.
My goal is to extract a specific path without extract all data (large volume):

related to my previous code:
=> R::Root/O_MyComplexStruct.MySimpleArrayStruct[0].B

Results should be an array of int

**Reply by LiorBanai** [reply]
Reading specific fields should be very easy. You can look at the unit test project which has many different reads.

**Reply by LiorBanai** [reply]
If you have issues please fell free to contact me

**Reply by wlaroche** [reply]
Is it expected it is not possible to add attributes when creating an object  (WriteObject) ? why ?
because It is possible with a WriteCompounds...

**Reply by LiorBanai** [reply]
hmm not really, could be missed case to handle. I'll check
can you add example could how you specifying the attributes?

**Reply by wlaroche** [reply]
Related to my first post, i must decode a complex structure (with struct arrays, multidimensional arrays...) from a hdf5 file (creating from an external tool in other language), and I'm still looking for the correct implementation in C# to read the data.... I have seen your samples but currently, nothing is working. I try some ways to know if i could read my data from a ReadObject or ReadCompound.

My h5 file contains some attributes, therefore i ask you if ReadObject could be the correct entry point but i don't think so.

**Reply by LiorBanai** [reply]
for attributes there are specific methods to deal with them (readAttributeXX). the Standard reads are for the objects themself 

btw if you are only needing read files I recommend https://github.com/Apollo3zehn/HDF5.NET as he has pure C# implementation
